src/libraries/Common/src/System/Security/Cryptography/MLKem.cs

```diff
@@ -1740,27 +1763,16 @@ private static void MLKemKeyReader(
             }
         }
 
-        private static void ThrowIfTrailingData(ReadOnlySpan<byte> data)
+        private static void ThrowIfMalformedEncoding(ReadOnlySpan<byte> data, AsnEncodingRules encoding)
         {
-            AsnDecoder.ReadEncodedValue(data, AsnEncodingRules.BER, out _, out _, out int bytesRead);
-
-            if (bytesRead != data.Length)
+            // TODO should we use ReadEncodedValue with try/catch instead so we can rethrow with a useful inner exception?
+            if (!AsnDecoder.TryReadEncodedValue(data, encoding, out _, out _, out _, out int bytesRead) || bytesRead != data.Length)
```

Comment:
This API does not "deeply" validate encoding rules, so I am not sure with the intention of that change.

That is to say: if you want to validate that an ASN.1 object is DER, this will not thrown for an indefinite length encoded item inside a definite length encoded item.

The actual reading will validate the rules - this was just meant to make sure there was nothing trailing after the data.

Reply:
My though process here was that in the previous code, `ReadEncodedValue` would throw `AsnContentException` (for example, when the length in the encoding was longer that the encoding itself), so either we can catch it and wrap it in a CryptoException or ArgumentException, or change it to the `TryReadEncodedValue` and throw. I did the latter, but if we're already doing the work of the check (even if it is non-exhaustive), it might make sense to surface it to the user as an inner exception. We can of course just let it throw the `AsnContentException` up too but I think the pattern in all the APIs is to wrap it in a CryptoException

Reply:
It does validate that the length field encoded in the input is not larger than the input itself (not enough data). And in the DER case it does validate that it isn't using variable length encoding. As you mentioned, these are both only for the top level structure, but so is the trailing data check.

How about `ThrowIfInvalidLength` or `ThrowIfUnexpectedLength`?

Reply:
> It does validate that the length field encoded in the input is not larger than the input itself (not enough data)

The `AsnReader` (or `AsnValueReader`) is going to catch that anyway when it attempts to actually decode the structure. Why try to pre-validate it? The only reason we have the trailing data check is is because it doesn't happen automatically during reading.

Reply:
> The AsnReader (or AsnValueReader) is going to catch that anyway

I believe that actually is the problem.  AsnDecoder.ReadEncodedValue threw an AsnContentException, this method didn't catch it, and an undocumented exception type escaped the public method.

Using TryRead avoids that exception, vs making a ThrowIf have a try/catch-wrap.

Reply:
> Using TryRead avoids that exception, vs making a ThrowIf have a try/catch-wrap.

Sure. My only point is that

1. The method does not actually validate the encoding in a meaningful way
2. It doesn't matter if it is BER/DER in this case.
3. It was already throwing crypto exceptions for truncated ASN.1.

Reply:
> it was already throwing crypto exceptions for truncated ASN.1.

If that's the case, then, full-undo.  If the edge case of the AsnContentException leaking out from the top-level item having a length that exceeds the span is true, then this is fine.

Reply:
> 3. It was already throwing crypto exceptions for truncated ASN.1.

It throws `AsnContentException`: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnDecoder.cs#L539-L540

> 2. It doesn't matter if it is BER/DER in this case.

Yes, I can relax the encoding to always be BER, but we still are going to have to wrap because of the comment above.